### PR TITLE
Integrates pqcrystals with common-aes / extends common code AES CTR-API

### LINKED ISSUE #973: Investigate integrating pqcrystals common code to OQS common code
as per [this discussion](https://github.com/open-quantum-safe/liboqs/pull/873#issuecomment-821557331)

### PR DESCRIPTION
**Pull request by bhess**
- Integrates pqcrystals (ref and avx2) with liboqs common AES code (for Kyber-90s and Dilithium-AES).
- Extends libOQS AES-CTR with Incremental API: `OQS_AES256_CTR_inc_init`, `OQS_AES256_CTR_inc_iv`, `OQS_AES256_CTR_inc_ivu64`, `OQS_AES256_CTR_inc_stream_iv`, `OQS_AES256_CTR_inc_stream_blks`.
- Adds some AES-CTR shim API.
- Faster AESNI CTR code (improved iv handling, 4x interleaved blocks for higher throughput).
- OpenSSL AES supporting CTR API.
- Updated pqcrystals patches (API, context releases).
- Removes redundant AES implementations from Kyber and Dilithium.
- Copy-from-upstream with updated patches.
- Uses internal AES on x86_64 (dist & AES) because of increased performance with the internal AESNI code.
- Adds AES-CTR benchmarks to speed_common

Fixes #973 

* [ ] Does this PR change the input/output behaviour of a cryptographic algorithm (i.e., does it change known answer test values)?  (If so, a version bump will be required from *x.y.z* to *x.(y+1).0*.)
* [ ] Does this PR change the the list of algorithms available -- either adding, removing, or renaming?  (If so, PRs in OQS-OpenSSL, OQS-BoringSSL, and OQS-OpenSSH will also be required by the time this is merged.)

### CODE REVIEW

File: src/common/pqclean_shims/aes.h [hunk 1]
```diff
@@ -17,7 +17,23 @@ typedef void *aes256ctx;
 #define aes256_ecb_keyexp(r, key) OQS_AES256_ECB_load_schedule((key), (r));
 #define aes256_ecb(out, in, nblocks, ctx) OQS_AES256_ECB_enc_sch((in), (nblocks) * AES_BLOCKBYTES, *(ctx), (out));
 #define aes256_ctr_keyexp(r, key) OQS_AES256_CTR_load_schedule((key), (r));
-#define aes256_ctr(out, outlen, iv, ctx) OQS_AES256_CTR_sch((iv), AESCTR_NONCEBYTES, *(ctx), (out), (outlen))
+#define aes256_ctr(out, outlen, iv, ctx) OQS_AES256_CTR_sch((iv), AESCTR_NONCEBYTES, *(ctx), (out), (outlen));
```
[the remark this answers is not in the document]
Comment (baentsch) [reply src/common/pqclean_shims/aes.h:20]: Sorry, I don't.

File: src/common/aes/aes.h [hunk 2]
```diff
@@ -95,19 +107,30 @@ void OQS_AES256_ECB_enc(const uint8_t *plaintext, const size_t plaintext_len, co
...
+ * AES counter mode keystream generator. A context generated by
+ * OQS_AES256_CTR_load_schedule() and OQS_AES256_CTR_load_nonce() is passed
+ * rather than a key and an IV.
+ *
+ * @param ctx      Abstract data structure for key schedule and IV.
  * @param out      Pointer to a block of memory which is big enough to contain out_len bytes; the result will be written here.
  * @param out_len  Length of output bytes to generate.
  */
-void OQS_AES256_CTR_sch(const uint8_t *iv, size_t iv_len, const void *schedule, uint8_t *out, size_t out_len);
+void OQS_AES256_CTR_sch_ivinit(void *ctx, uint8_t *out, size_t out_len);
```
Comment (dstebila) [inline src/common/aes/aes.h:133]: Why is this called `OQS_AES256_CTR_sch_ivinit`? The `ivinit` part of it makes it sound like it's going to initialize the IV, but in fact it generates output...

Comment (bhess) [reply src/common/aes/aes.h]: Right, it's actually an update function since it internally updates the IV/counter and allows to be called multiple times. I've now renamed it to `OQS_AES256_CTR_sch_upd_blks`.

**Comment by bhess** [comment]
The internal AESNI implementation is now quite a bit faster than OpenSSL. This is noticeable mainly in Kyber-90s and Dilithium-AES, but also in NTRULPR and Classic McEliece. The PR therefore turns off OpenSSL AES by default on x86_64 (dist and AESNI builds).

```
Configuration info
==================
Target platform:  x86_64-Darwin-21.5.0
Compiler:         clang (13.1.6 (clang-1316.0.21.2.5))
Compile options:  [-march=native;-Werror;-Wall;-Wextra;-Wpedantic;-Wno-unused-command-line-argument;-O3;-fomit-frame-pointer;-Wbad-function-cast;-Wcast-qual;-Wnarrowing;-Wconversion]
OQS version:      0.7.2-dev
Git commit:       773b547a2a2539b8d1329a5feb013a06d3e427b4 (+ local modifications)
OpenSSL enabled:  Yes (OpenSSL 1.1.1n  15 Mar 2022)
AES:              NI
SHA-2:            OpenSSL
SHA-3:            C
OQS build flags:  OQS_OPT_TARGET=auto CMAKE_BUILD_TYPE=Release 
CPU exts compile-time:  ADX AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3

Speed test
==========
Started at 2022-06-16 09:24:20
Operation                            | Iterations | Total time (s) | Time (us): mean | pop. stdev | CPU cycles: mean          | pop. stdev
------------------------------------ | ----------:| --------------:| ---------------:| ----------:| -------------------------:| ----------:
OQS_AES128_ECB_load+free_sch         |    5275205 |          1.000 |           0.190 |      0.543 |                       396 |        219
OQS_AES128_ECB_enc_sch               |   15559558 |          1.000 |           0.064 |      0.746 |                        42 |        298
OQS_AES128_ECB_enc                   |    4874984 |          1.000 |           0.205 |      0.505 |                       437 |        135
OQS_AES256_ECB_load+free_sch         |    4772176 |          1.000 |           0.210 |      2.920 |                       448 |       7472
OQS_AES256_ECB_enc_sch               |   13279853 |          1.000 |           0.075 |      0.267 |                        58 |         74
OQS_AES256_ECB_enc                   |    4263366 |          1.000 |           0.235 |      0.432 |                       513 |        188
OQS_AES256_CTR_init+iv+free          |    4781832 |          1.000 |           0.209 |      0.507 |                       447 |        162
OQS_AES256_CTR_inc_stream_iv         |   15052486 |          1.000 |           0.066 |      0.932 |                        44 |       2053
OQS_AES256_CTR_inc_stream_blks       |   15339421 |          1.000 |           0.065 |      0.321 |                        44 |        525
```

```
Configuration info
==================
Target platform:  x86_64-Darwin-21.5.0
Compiler:         clang (13.1.6 (clang-1316.0.21.2.5))
Compile options:  [-march=native;-Werror;-Wall;-Wextra;-Wpedantic;-Wno-unused-command-line-argument;-O3;-fomit-frame-pointer;-Wbad-function-cast;-Wcast-qual;-Wnarrowing;-Wconversion]
OQS version:      0.7.2-dev
Git commit:       773b547a2a2539b8d1329a5feb013a06d3e427b4 (+ local modifications)
OpenSSL enabled:  Yes (OpenSSL 1.1.1n  15 Mar 2022)
AES:              OpenSSL
SHA-2:            OpenSSL
SHA-3:            C
OQS build flags:  OQS_OPT_TARGET=auto CMAKE_BUILD_TYPE=Release 
CPU exts compile-time:  ADX AES AVX AVX2 BMI1 BMI2 PCLMULQDQ POPCNT SSE SSE2 SSE3

Speed test
==========
Started at 2022-06-16 09:57:12
Operation                            | Iterations | Total time (s) | Time (us): mean | pop. stdev | CPU cycles: mean          | pop. stdev
------------------------------------ | ----------:| --------------:| ---------------:| ----------:| -------------------------:| ----------:
OQS_AES128_ECB_load+free_sch         |    1986331 |          1.000 |           0.503 |      0.507 |                      1209 |        209
OQS_AES128_ECB_enc_sch               |   12281066 |          1.000 |           0.081 |      0.285 |                       106 |         67
OQS_AES128_ECB_enc                   |    1766067 |          1.000 |           0.566 |      9.189 |                      1372 |      23778
OQS_AES256_ECB_load+free_sch         |    1949958 |          1.000 |           0.513 |      0.603 |                      1232 |        555
OQS_AES256_ECB_enc_sch               |   11175312 |          1.000 |           0.089 |      2.196 |                       119 |       5411
OQS_AES256_ECB_enc                   |    1798036 |          1.000 |           0.556 |      0.506 |                      1347 |        236
OQS_AES256_CTR_init+iv+free          |    1927037 |          1.000 |           0.519 |      0.827 |                      1248 |       1691
OQS_AES256_CTR_inc_stream_iv         |    2012398 |          1.000 |           0.497 |      0.978 |                      1192 |       2170
OQS_AES256_CTR_inc_stream_blks       |    9790800 |          1.000 |           0.102 |      0.307 |                       148 |        114
```

File: src/common/aes/aes.h (hunk 2 above)
Comment (bhess) [reply src/common/aes/aes.h]: The API naming is now more in line with the "incremental" API from SHA2/3. (see PR comments)

File: .CMake/alg_support.cmake [hunk 3]
```diff
@@ -15,7 +15,10 @@ if(NOT WIN32)
 else()
     option(OQS_USE_OPENSSL "Enable OpenSSL usage" OFF)
 endif()
-cmake_dependent_option(OQS_USE_AES_OPENSSL "" ON "OQS_USE_OPENSSL" OFF)
+
+# Use OpenSSL's AES only if no AESNI and x86 dist build is used.
+# Reason: AESNI implementation better fits our incremental API.
+cmake_dependent_option(OQS_USE_AES_OPENSSL "" ON "OQS_USE_OPENSSL; NOT OQS_DIST_X86_64_BUILD; NOT OQS_USE_AES_INSTRUCTIONS" OFF)
```
Comment (baentsch) [inline .CMake/alg_support.cmake:21]: I'm at a loss understanding this code: What defines exactly need to be set to disable AESNI? See #1229 to document this (as and when merged).

Comment (bhess) [reply .CMake/alg_support.cmake:21]: > See #1229 to document this (as and when merged).

thx, will add documentation once #1229 is merged.

Comment (bhess) [reply .CMake/alg_support.cmake:21]: To test different variants I was directly editing alg_support.cmake.

Looks like turning on/off `OQS_USE_${CPU_EXTENSION}_INSTRUCTIONS` via command line options will be overridden by the code here:
https://github.com/open-quantum-safe/liboqs/blob/e37ad5b5d8a0c7cc19a8a148c7adce710a010b6b/.CMake/gcc_clang_intrinsics.cmake#L14-L16
It should probably guarded like this:

```
foreach(CPU_EXTENSION ${RUN_OUTPUT})
    if (NOT DEFINED OQS_USE_${CPU_EXTENSION}_INSTRUCTIONS)
        set(OQS_USE_${CPU_EXTENSION}_INSTRUCTIONS ON)
     endif()
endforeach()
```

Comment (baentsch) [reply .CMake/alg_support.cmake:21]: > added a commit with this change

Thanks!

File: src/kem/kyber/pqcrystals-kyber_kyber1024-90s_ref/symmetric-aes.c [hunk 4]
```diff
@@ -6,10 +6,11 @@
 
 void kyber_aes256xof_absorb(aes256ctr_ctx *state, const uint8_t seed[32], uint8_t x, uint8_t y)
 {
+  (void) seed;
```
Comment (dstebila) [inline src/kem/kyber/pqcrystals-kyber_kyber1024-90s_ref/symmetric-aes.c:9]: Is it okay that seed is never used? Does this not change the functionality?

Comment (dstebila) [reply src/kem/kyber/pqcrystals-kyber_kyber1024-90s_ref/symmetric-aes.c:9]: Oh, I see this has been incorporated into `xof_init`.

**Review by dstebila — APPROVED** [review]

File: CONFIGURE.md [hunk 5]
```diff
@@ -59,7 +59,7 @@ When built for use on a single machine, the library will only include the best a
 
 ## OQS_USE_OPENSSL
 
-This can be set to `ON` or `OFF`. When `ON`, the additional options `OQS_USE_AES_OPENSSL`, `OQS_USE_SHA2_OPENSSL`, and `OQS_USE_SHA3_OPENSSL` are made available to control whether liboqs uses OpenSSL's AES, SHA-2, and SHA-3 implementations. By default, `OQS_USE_AES_OPENSSL` and `OQS_USE_SHA2_OPENSSL` are `ON` while `OQS_USE_SHA3_OPENSSL` is `OFF`.
+This can be set to `ON` or `OFF`. When `ON`, the additional options `OQS_USE_AES_OPENSSL`, `OQS_USE_SHA2_OPENSSL`, and `OQS_USE_SHA3_OPENSSL` are made available to control whether liboqs uses OpenSSL's AES, SHA-2, and SHA-3 implementations. By default, `OQS_USE_AES_OPENSSL` is `ON` if `OQS_DIST_X86_64_BUILD` and `OQS_USE_AES_INSTRUCTIONS` are not set, `OQS_USE_SHA2_OPENSSL` is `ON` while `OQS_USE_SHA3_OPENSSL` is `OFF`.
```
Comment (baentsch) [inline CONFIGURE.md:62]: Can this be facilitated/explained using `OQS_DIST_BUILD` (documented variable) instead of `OQS_DIST_X86_64_BUILD` (i.e., what happens on ARM -- where `OQS_DIST_X86_64_BUILD` surely is not set)? Otherwise, there should be documentation for `OQS_DIST_X86_64_BUILD`. `OQS_USE_AES_INSTRUCTIONS` also is not (yet) documented in [CONFIGURE.md](https://github.com/open-quantum-safe/liboqs/blob/main/CONFIGURE.md).

File: scripts/copy_from_upstream/copy_from_upstream.yml [hunk 6]
```diff
@@ -15,18 +15,18 @@ upstreams:
     git_branch: master
     git_commit: faf5c3fe33e0b61c7c8a7888dd862bf5def17ad2
     kem_meta_path: '{pretty_name_full}_META.yml'
-    common_meta_path: 'Common_META.yml'
+    #common_meta_path: 'Common_META.yml'
```
Comment (baentsch) [inline scripts/copy_from_upstream/copy_from_upstream.yml:18]: Is this not needed any more? If so, why not delete?

File: src/common/aes/aes.h [hunk 7]
```diff
@@ -54,25 +54,48 @@ void OQS_AES128_ECB_enc_sch(const uint8_t *plaintext, const size_t plaintext_len
...
+ * bytes are initialized to all zeros.
+ *
+ * @param iv             Initialization Vector.
+ * @param iv_len         Length of the initialization vector.
+ * @param ctx            Abstract data structure for IV.
+ */
+void OQS_AES256_CTR_inc_iv(const uint8_t *iv, size_t iv_len, void *ctx);
+
+/**
+ * Function to fill a context fiven an IV for use in CTR mode.
```
Comment (baentsch) [inline src/common/aes/aes.h:83]: typo "fiven". In general the question: What does "inc" stand for in the function name?

File: src/common/aes/aes.h [hunk 8]
```diff
@@ -54,25 +54,48 @@ void OQS_AES128_ECB_enc_sch(const uint8_t *plaintext, const size_t plaintext_len
...
+ * Handles an 8-byte IV passed as a 64-bit unsigned integer,
+ * counter bytes are initialized to zero.
+ *
+ * @param iv             Initialization Vector as 64-bit integer.
+ * @param ctx            Abstract datra structure for IV.
```
Comment (baentsch) [inline src/common/aes/aes.h:88]: typo "datra". Same question as above about function name.

File: CONFIGURE.md (hunk 5 above)
Comment (bhess) [reply CONFIGURE.md:62]: I'd rephrase it to "`OQS_DIST_BUILD` on x86_64" and add a section describing the options to (un)set CPU features.

File: src/common/aes/aes_ossl.c [hunk 9]
```diff
@@ -65,15 +77,42 @@ void OQS_AES256_ECB_load_schedule(const uint8_t *key, void **schedule) {
...
 }
 
-void OQS_AES256_CTR_load_schedule(const uint8_t *key, void **schedule) {
+void OQS_AES256_CTR_inc_init(const uint8_t *key, void **schedule) {
 	*schedule = malloc(sizeof(struct key_schedule));
-	OQS_EXIT_IF_NULLPTR(*schedule);
 	struct key_schedule *ks = (struct key_schedule *) *schedule;
+	EVP_CIPHER_CTX *ctr_ctx = EVP_CIPHER_CTX_new();
+	assert(ctr_ctx != NULL);
+	//OQS_OPENSSL_GUARD(EVP_EncryptInit_ex(ctr_ctx, EVP_aes_256_ctr(), NULL, key, NULL));
```
Comment (baentsch) [inline src/common/aes/aes_ossl.c:85]: commented guard apparently initializing OpenSSL function intentional?

Comment (bhess) [reply src/common/aes/aes_ossl.c:85]: yes, will remove.

File: CONFIGURE.md (hunk 5 above)
Comment (bhess) [reply CONFIGURE.md]: (updated in 0b91245833c1bb9756519316fb13b8cb4d67d2cd)

File: scripts/copy_from_upstream/copy_from_upstream.yml (hunk 6 above)
Comment (bhess) [reply scripts/copy_from_upstream/copy_from_upstream.yml]: updated in 0b91245833c1bb9756519316fb13b8cb4d67d2cd

File: src/common/aes/aes.h (hunk 7 above)
Comment (bhess) [reply src/common/aes/aes.h]: typo fixed in 0b91245833c1bb9756519316fb13b8cb4d67d2cd

File: src/common/aes/aes_ossl.c (hunk 9 above)
Comment (bhess) [reply src/common/aes/aes_ossl.c]: updated in 0b91245833c1bb9756519316fb13b8cb4d67d2cd

**Review by baentsch — APPROVED** [review]
Thanks for this common code improvement! Thanks also for the various updates as per feedback. Now LGTM. 
Related question (to anyone reading): Would it be reasonable to add the (particularly common code) performance testing (as now also added for AES CTR) to CI to guard against regressions?

**Comment by bhess** [comment]
> Thanks for this common code improvement! Thanks also for the various updates as per feedback. Now LGTM.
> Related question (to anyone reading): Would it be reasonable to add the (particularly common code) performance testing (as now also added for AES CTR) to CI to guard against regressions?

Thanks for the review! Yes I think it would be reasonable to add some test for the common code. Either as part of CI or as part of profiling.
